src/Common/KnownObjectNames.cpp

```diff
@@ -0,0 +1,48 @@
+#include <Common/KnownObjectNames.h>
+#include <Poco/String.h>
+
+
+namespace DB
+{
+
+bool KnownObjectNames::exists(const String & name) const
+{
+    std::lock_guard lock{mutex};
+    if (names.contains(name))
+        return true;
+
+    if (!case_insensitive_names.empty())
+    {
+        String lower_name = Poco::toLower(name);
+        if (case_insensitive_names.contains(lower_name))
+            return true;
+    }
+
+    return false;
+}
+
+
+void KnownObjectNames::add(const String & name, bool case_insensitive)
+{
+    std::lock_guard lock{mutex};
+    if (case_insensitive)
+        case_insensitive_names.emplace(Poco::toLower(name));
+    else
+        names.emplace(name);
+}
+
+
+KnownTableFunctionNames & KnownTableFunctionNames::instance()
```

Comment:
I think we can use TableFunctionFactory & FormatFactory to check if a function/format name exists. Since we already have all the information there, might be easier to add a function there like contains() instead of storing same information in a set and having another singleton. What do you think?

Reply:
It's because of dependencies. `ASTFunction` is in the `clickhouse_parsers` target, and `TableFunctionFactory` is in the `clickhouse_table_functions` target. `clickhouse_table_functions` is a big target and it depends on `clickhouse_parsers`.
However to be able to address `TableFunctionFactory` `clickhouse_parsers` must depend on `clickhouse_table_functions`. So I've introduced this singleton to overcome this situation.

Reply:
okay, got it